libtenzir/builtins/operators/compress_decompress.cpp

```diff
@@ -0,0 +1,387 @@
+      .pretty_name("operator_args")
+      .fields(f.field("type", x.type), f.field("level", x.level));
+  }
+};
+
+auto codec_from_args(const operator_args& args)
+  -> arrow::Result<std::shared_ptr<arrow::util::Codec>> {
+  auto compression_type
+    = arrow::util::Codec::GetCompressionType(args.type.inner);
+  // Arrow straight up crashes if we use a codec created from the string
+  // "uncompressed", so we just don't do that. Last checked with Arrow 12.0.
+  if (not compression_type.ok()
+      or *compression_type == arrow::Compression::UNCOMPRESSED) {
+    return arrow::Status::Invalid(
+      fmt::format("failed to get compression type `{}`; must be one of "
+                  "`brotli`, `bz2`, `gzip`, `lz4`, `zstd`",
+                  args.type.inner));
+  }
+  const auto compression_level
+    = args.level ? args.level->inner : arrow::util::kUseDefaultCompressionLevel;
+  auto codec = arrow::util::Codec::Create(*compression_type, compression_level);
+  return codec;
+}
+
+class compress_operator final : public crtp_operator<compress_operator> {
+public:
+  compress_operator() = default;
+
+  explicit compress_operator(operator_args args) : args_{std::move(args)} {
+  }
+
+  auto operator()(generator<chunk_ptr> input,
+                  operator_control_plane& ctrl) const -> generator<chunk_ptr> {
+    auto codec = codec_from_args(args_);
+    if (not codec.ok()) {
+      diagnostic::error("failed to create codec for compression type `{}`: {}",
+                        args_.type.inner, codec.status().ToString())
+        .primary(args_.type.source)
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+    auto compressor = codec.ValueUnsafe()->MakeCompressor();
+    if (not compressor.ok()) {
+      diagnostic::error("failed to create compressor: {}",
+                        compressor.status().ToString())
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+    auto out_buffer = std::vector<uint8_t>{};
+    out_buffer.resize(1 << 20);
+    auto in_buffer = input_buffer{};
+    for (auto&& bytes : input) {
+      if (not bytes) {
+        co_yield {};
+        continue;
+      }
+      in_buffer.consume(std::move(bytes));
+      while (in_buffer.size() > 0) {
+        auto result = compressor.ValueUnsafe()->Compress(
+          in_buffer.size(), in_buffer.data(),
+          detail::narrow_cast<int64_t>(out_buffer.size()), out_buffer.data());
+        if (not result.ok()) {
+          diagnostic::error("failed to compress: {}",
+                            result.status().ToString())
+            .emit(ctrl.diagnostics());
+          co_return;
+        }
+        // Some compressors signal that they need a larger output buffer to
+        // write into. We already start with a rather large one, but in case
+        // that isn't enough we just double the size until the compressor
+        // stops complaining.
+        if (result->bytes_read == 0) [[unlikely]] {
+          out_buffer.resize(out_buffer.size() * 2);
```

Comment:
`max_size()` check for the new vector size here? (Although it may be very unlikely)

Reply:
Good call!